stan/math/prim/functor/apply_vector_unary.hpp

```diff
@@ -48,13 +43,38 @@ struct apply_vector_unary<T, require_eigen_t<T>> {
   template <typename F, typename T2 = T,
             require_t<is_eigen_matrix_base<plain_type_t<T2>>>* = nullptr>
   static inline auto apply(const T& x, const F& f) {
-    return f(x).matrix().eval();
+    return make_holder([](const auto& a) { return a.matrix().derived(); },
+                       f(x));
   }
 
   template <typename F, typename T2 = T,
             require_t<is_eigen_array<plain_type_t<T2>>>* = nullptr>
   static inline auto apply(const T& x, const F& f) {
-    return f(x).array().eval();
+    return make_holder([](const auto& a) { return a.array().derived(); }, f(x));
+  }
+
+  /**
+   * Member function for applying a functor to a vector and subsequently
+   * returning a vector. This is a variant of `apply` that does not construct
+   * `holder`, so it is up to the caller to ensure the returned expression is
+   * evaluated before `x` is destructed.
+   *
+   * @tparam T Type of argument to which functor is applied.
+   * @tparam F Type of functor to apply.
+   * @param x Eigen input to which operation is applied.
+   * @param f functor to apply to Eigen input.
+   * @return Eigen object with result of applying functor to input
+   */
+  template <typename F, typename T2 = T,
+            require_t<is_eigen_matrix_base<plain_type_t<T2>>>* = nullptr>
+  static inline auto apply_no_holder(const T& x, const F& f) {
+    return f(x).matrix().derived();
```

Comment:
I missed the `.derived()` before. Why `.derived()`?

Reply:
Eigen::MatrixXd().matrix() returns some wrapper class. By calling `derived()` we get rid of it and can return the actual expression/matrix type.